docs: add turtlebot4 tutorial (#240)
RobotecAI/rai: src/examples/turtlebot4/README.md

# Turtlebot4 tutorial

Following tutorial will be presented during [ROSCon 2024](https://roscon.ros.org/2024/) talk about RAI.

## Recording

TBD

## Step by step instructions

1. O3DE simulation setup

   1. Download turtlebot4 simulation binary matching your platform from here (TODO link).

      > **NOTE** If you would like to make changes to the simulation and create your own binary please follow this [README.md](https://github.com/RobotecAI/ROSCon2024Tutorial/README.md)

   2. Unzip it:

      ```bash
      cd rai
      # for Ubuntu 22.04 jammy and ros2 humlbe

>       unzip -d src/examples/turtlebot4/simulation Turtlebot4_PR1_jammyhumlbe.zip

Reviewer: ```suggestion
      unzip -d src/examples/turtlebot4/simulation Turtlebot4_PR1_jammyhumble.zip
```

Prolly a typo?

# Turtlebot4 tutorial

> The following tutorial will be presented during [ROSCon 2024](https://roscon.ros.org/2024/)

Reviewer 2: This is less interesting information as the tutorial is to have longevity beyond ROSCon, and once guided here, people will already know.

Author: fixed

talk about RAI.

## Recording

TBD

## Step by step instructions

1. O3DE simulation setup

   1. Download turtlebot4 simulation binary matching your platform from :

      - Ubuntu 22.04 & ros2 humble: [link](https://robotecai-my.sharepoint.com/:u:/g/personal/bartlomiej_boczek_robotec_ai/EZLmGtPNgl9Kiu4royJJnVgB5tjS2Vze0myXDyVJtNcnRw?e=L42Z4z)
      - Ubuntu 24.04 & ros2 jazzy: [link](https://robotecai-my.sharepoint.com/:u:/g/personal/bartlomiej_boczek_robotec_ai/Ebi0O3eWAtRAl3GL8LuEBqIB0-5fnHDCm7yOtcTucpNfzA?e=G6eJKP)

      > **NOTE** If you would like to make changes to the simulation and create your
      > own binary please follow this [README.md](https://github.com/RobotecAI/ROSCon2024Tutorial/README.md)

   2. Unzip it:

      ```bash
      cd rai

>       # for Ubuntu 22.04 jammy and ros2 humlbe

Reviewer: `cd rai` before rai is cloned

Author: done

      unzip -d src/examples/turtlebot4/simulation Turtlebot4_jammyhumble_0.0.1.zip

      # for Ubuntu 24.04 noble and ros2 jazzy
      unzip -d src/examples/turtlebot4/simulation Turtlebot4_noblejazzy_0.0.1.zip
      ```

2. Clone and install [rai](https://github.com/RobotecAI/rai)

>    1. Follow setps from [1. Setting up the workspace](https://github.com/RobotecAI/rai?tab=readme-ov-file#1-setting-up-the-workspace)

Reviewer 2: ```suggestion
   1. Follow steps from [1. Setting up the workspace](https://github.com/RobotecAI/rai?tab=readme-ov-file#1-setting-up-the-workspace)
```

   2. Then follow steps from [2. Build the project](https://github.com/RobotecAI/rai?tab=readme-ov-file#2-build-the-project)

> 3. Setup your [OpenAI API](https://openai.com/index/openai-api/). Make sure you have set `OPENAI_API_KEY` variable.

Reviewer 2: Instead, link to RAI vendor-agnostic setup. You can summarize (e.g. follow steps 1-3 of RAI)

Author: unfortunately this tutorial will not work well with Ollama, because of lack of vision-function calling model..

4. Configure `rai_whoami_node` (based on ["Your robot identity in RAI"](https://github.com/RobotecAI/rai/blob/development/docs/create_robots_whoami.md) tutorial):

   1. Create `whoami` package for turtlebot4 in `src/examples/turtlebot4`

      ```bash
      . ./setup_shell.sh
      poetry run create_rai_ws --name turtlebot4 --destination-directory src/examples
      ```

   2. Download official turtlebot4 [data sheet](https://bit.ly/3KCp3Du) into
      `src/examples/turtlebot4_whoami/description/documentation`
   3. Download [image](https://s3.amazonaws.com/assets.clearpathrobotics.com/wp-content/uploads/2022/03/16113604/Turtlebot-4-20220207.44.png) of turtlebot4 into `src/examples/turtlebot4_whoami/description/images`
   4. Create robot's identity. Run the `parse_whoami_package`. This will process the documentation, building
      it into a vector database, which is used by RAI agent to reason about its identity.

>       > **NOTE**: Vector database is created using the OpenAI API. Parsing bigger documents

Reviewer 2: Is this possible with another vendor? 
Provide alternative of vector db download instead of building.

Author: @adamdbrw Ollama can be used to do embeddings, openai will be used to summarize the image. We could summarize the image with LLava model, but with current config setup user would have to change the model before next step (llava doens't support tool calling) - therefore I would skip it not to overcomplicate

      > might lead to costs. Embedding model can be configured in
      > [config.toml](https://github.com/RobotecAI/rai/blob/development/config.toml#L13)

      ```bash
      poetry run parse_whoami_package src/examples/turtlebot4_whoami/description
      # you will be asked to press `y` to continue
      ```

>    5. Rebuild the workspace

Reviewer 2: ```suggestion
   5. Build the workspace which now includes the new package
```

      ```bash
      colcon build --symlink-install
      ```

>    Congratulations! Your `rai_whoami_node` is configured. In the follwing steps

Reviewer 2: ```suggestion
   Congratulations! Your `rai_whoami_node` is configured. In the following steps
```

Reviewer 2: Also please double check spelling overall etc.

>    your RAI agent will have turtlebot4 personality.

Reviewer 2: ```suggestion
   your RAI agent will assume a turtlebot4 "personality".
```

5. Run rai agent:

   ```bash
   ros2 launch ./src/examples/turtlebot4/turtlebot.launch.xml \

>        game_launcher:=./src/examples/turtlebot4/simulation/TurtleBot4DemoGamePackage/TurtleBot4Demo.GameLauncher

Reviewer: Ubuntu 22 humble fails with:

```
[rai_whoami_node-1] [INFO] [1729006016.012198943] [rai_whoami_node]: Robot constitution loaded from /home/krachwal/Projects/internal/rai/install/turtlebot4_whoami/share/turtlebot4_whoami/description/robot_constitution.txt
[rai_whoami_node-1] Traceback (most recent call last):
[rai_whoami_node-1]   File "/home/krachwal/Projects/internal/rai/install/rai_whoami/lib/rai_whoami/rai_whoami_node", line 33, in <module>
[rai_whoami_node-1]     sys.exit(load_entry_point('rai-whoami', 'console_scripts', 'rai_whoami_node')())
[rai_whoami_node-1]   File "/home/krachwal/Projects/internal/rai/build/rai_whoami/rai_whoami/rai_whoami_node.py", line 191, in main
[rai_whoami_node-1]     who_am_i_node = WhoAmI()
[rai_whoami_node-1]   File "/home/krachwal/Projects/internal/rai/build/rai_whoami/rai_whoami/rai_whoami_node.py", line 81, in __init__
[rai_whoami_node-1]     self.faiss_index = self._load_documentation()
[rai_whoami_node-1]   File "/home/krachwal/Projects/internal/rai/build/rai_whoami/rai_whoami/rai_whoami_node.py", line 84, in _load_documentation
[rai_whoami_node-1]     faiss_index = FAISS.load_local(
[rai_whoami_node-1]   File "/home/krachwal/.cache/pypoetry/virtualenvs/rai-Qp9irBSo-py3.10/lib/python3.10/site-packages/langchain_community/vectorstores/faiss.py", line 1203, in load_local
[rai_whoami_node-1]     index = faiss.read_index(str(path / f"{index_name}.faiss"))
[rai_whoami_node-1]   File "/home/krachwal/.cache/pypoetry/virtualenvs/rai-Qp9irBSo-py3.10/lib/python3.10/site-packages/faiss/swigfaiss_avx2.py", line 10538, in read_index
[rai_whoami_node-1]     return _swigfaiss_avx2.read_index(*args)
[rai_whoami_node-1] RuntimeError: Error in faiss::FileIOReader::FileIOReader(const char*) at /project/faiss/faiss/impl/io.cpp:67: Error: 'f' failed: could not open /home/krachwal/Projects/internal/rai/install/turtlebot4_whoami/share/turtlebot4_whoami/description/index.faiss for reading: No such file or directory
[ERROR] [rai_whoami_node-1]: process has died [pid 58662, exit code 1, cmd '/home/krachwal/Projects/internal/rai/install/rai_whoami/lib/rai_whoami/rai_whoami_node --ros-args -r __node:=rai_whoami_node --params-file /tmp/launch_params_r1_cikmj'].
```

Author: I rebased the branch on the newest changes in whoami from development - please try again

   ```

> 6. Open you internet browser and go to `localhost:8501`

Reviewer 2: ```suggestion
```

Reviewer 2: Note - explaining purpose of action in the next point

Author: fixed

> 7. You can interact with your RAI agent through the chat. On the left you can interact

Reviewer 2: ```suggestion
7. Open your internet browser with the address `localhost:8501` to can interact with your RAI agent through the chat. On the left you can communicate
```

   with the RAI HMI Agent. On the right you will see status of missions that were send
   to the RAI Node and are executed on your robot.

   - `HMI Agent` is responsible for human-robot interaction.
   - `RAI Node` is responsible for executing tasks on the robot.

>    For example you can try such prompt:

Reviewer 2: ```suggestion
   For example you can try such prompts:
```

   - Are you able to bring you something from the kitchen? (testing robot's identity)
   - What are your ros2 interfaces? (discovery of ros2 interfaces)
   - tell me what you see (interpretation of camera image)
   - Drive towards the chair (when table is not visible, robot rejects task that it cannot do)
   - Spin yourself left by 45 degrees (interaction with the robot using ros2 interfaces) - table with the robotic are shoud be visible in the camera
   - Use robotic arm to pick up a box from the table (identity and intefaces doesn't allow it)
   - Drive towards the table (when table is visible, testing ability to interpret camera image and perform actions based on the knowledge)

>    - Drive behind the table (likely will fail in current `rai` version due to limitations in spatial reasoning)

Reviewer 2: ```suggestion
```

Reviewer 2: Remove the example that is likely to fail

### Troubleshooting

> _My robot doesn't have the identity._

Reviewer 2: ```suggestion
_My robot doesn't have an identity._
```

Reviewer 2: Also add how this would look like for the user, e.g seeing an error message etc.

1. Query `rai_whoami_node` for the identity manually

   ```bash
   ros2 service call /rai_whoami_identity_service std_srvs/srv/Trigger
   ```

> 2. Make sure you have all required files were generated correctly in the `turtlebot4_whoami` package and have simillar sizes.

Reviewer 2: ```suggestion
2. Make sure all required files were generated correctly in the `turtlebot4_whoami` package and have similar sizes to those listed below.
```

   ```bash
   ls -sh src/examples/turtlebot4_whoami/description/
   total 52K
   4.0K documentation  4.0K images   28K index.faiss  8.0K index.pkl  4.0K robot_constitution.txt  4.0K robot_identity.txt
   ```

   You can also check the contents of `robot_indentify.txt` file (it is generated by LLM, but should be simillar to the one below).

   ```bash
   cat src/examples/turtlebot4_whoami/description/robot_identity.txt
   I am a TurtleBot® 4, a robotics learning platform designed for education and
   research in robotics. This next-generation mobile robot is built on the iRobot®
   Create® 3 mobile base and comes fully assembled with ROS 2 installed and configured,
   making it accessible for beginners and experienced developers alike.
   ...
   ```

>    If files above are incorrect please check your `OPENAI_API_KEY` and rerun point 4

Reviewer 2: Vendor independent solution?

Author: https://github.com/RobotecAI/rai/pull/240/commits/f4d48d5a89c54637af3f103128e65970861cbb7f

   of this tutorial.

---

_Robot doesn't move in the simulation._

1. Make sure you can see ros2 topic when simulation binary is running

   ```bash
   ros2 topic list

>    ```

Reviewer: I think the name of the expected ros2 topic should be provided

Author: https://github.com/RobotecAI/rai/pull/240/commits/a908fa030b95b5e10935f7657624827fa35dad19

## Step by step instructions

1. Clone and install [rai](https://github.com/RobotecAI/rai)

   1. Follow steps from [1. Setting up the workspace](https://github.com/RobotecAI/rai?tab=readme-ov-file#1-setting-up-the-workspace)
   2. Then follow steps from [2. Build the project](https://github.com/RobotecAI/rai?tab=readme-ov-file#2-build-the-project)

2. O3DE simulation setup

   1. Download turtlebot4 simulation binary matching your platform from :

      - Ubuntu 22.04 & ros2 humble: [link](https://robotecai-my.sharepoint.com/:u:/g/personal/bartlomiej_boczek_robotec_ai/EZLmGtPNgl9Kiu4royJJnVgB5tjS2Vze0myXDyVJtNcnRw?e=L42Z4z)
      - Ubuntu 24.04 & ros2 jazzy: [link](https://robotecai-my.sharepoint.com/:u:/g/personal/bartlomiej_boczek_robotec_ai/ETkT-jvozlpBtuG1piDeqggBRmWl5eylIChc_g0v_EetpA?e=EjcJqe)

      > **NOTE** If you would like to make changes to the simulation and create your
      > own binary please follow this [README.md](https://github.com/RobotecAI/ROSCon2024Tutorial/README.md)

   2. Unzip it:

      ```bash
      cd rai

>       # for Ubuntu 22.04 jammy and ros2 humlbe

Reviewer 3: ```suggestion
      # for Ubuntu 22.04 jammy and ros2 humble
```

      > **NOTE**: Vector database is created using the OpenAI API. Parsing bigger documents
      > might lead to costs. Embedding model can be configured in
      > [config.toml](https://github.com/RobotecAI/rai/blob/development/config.toml#L13)

      ```bash
      poetry run parse_whoami_package src/examples/turtlebot4_whoami
      # you will be asked to press `y` to continue
      ```

   5. Build the workspace which now includes the new package

      ```bash
      deactivate # if poetry env is activated
      colcon build --symlink-install
      ```

   6. Ensure `turtlebot4_whoami` package has been built:

      ```bash
      . ./setup_shell.sh
      ros2 pkg list | grep turtlebot4
      ```

   Congratulations! Your `rai_whoami_node` is configured. In the following steps
   your RAI agent will assume a turtlebot4 "personality".

5. Run rai agent:

   ```bash
   ros2 launch ./src/examples/turtlebot4/turtlebot.launch.xml \
       game_launcher:=./src/examples/turtlebot4/simulation/TurtleBot4DemoGamePackage/TurtleBot4Demo.GameLauncher
   ```

   You should be able to see a simulation scene.

>    > **Tip**: Press 1, 2 or 3 on the keyboard when simulation window to change the

Reviewer 3: These shortcuts does not work for me - is this still available? WSAD camera control works fine.

Reviewer 4: It might be caused by a bug in 24.09. Which binary did you use?

## Step by step instructions

0. Install nav2 packates

   ```shell
   sudo apt install "ros-${ROS_DISTRO}-nav2-*"

   # add to ~/.bashrc
   . /opt/ros/humble/setup.bash
   # or
   . ./opt/ros/jazzy/setup.bash
   export RMW_IMPLEMENTATION=rmw_fastrtps_cpp

   ```

1. Clone and install [rai](https://github.com/RobotecAI/rai)

   1. Follow steps from [1. Setting up the workspace](https://github.com/RobotecAI/rai?tab=readme-ov-file#1-setting-up-the-workspace)
   2. Then follow steps from [2. Build the project](https://github.com/RobotecAI/rai?tab=readme-ov-file#2-build-the-project)

2. O3DE simulation setup

   1. Download turtlebot4 simulation binary matching your platform from :

      - Ubuntu 22.04 & ros2 humble: [link](https://robotecai-my.sharepoint.com/:u:/g/personal/bartlomiej_boczek_robotec_ai/EZLmGtPNgl9Kiu4royJJnVgB5tjS2Vze0myXDyVJtNcnRw?e=L42Z4z)
      - Ubuntu 24.04 & ros2 jazzy: [link](https://robotecai-my.sharepoint.com/:u:/g/personal/bartlomiej_boczek_robotec_ai/ETkT-jvozlpBtuG1piDeqggBRmWl5eylIChc_g0v_EetpA?e=EjcJqe)

>       > **NOTE** If you would like to make changes to the simulation and create your
>       > own binary please follow this [README.md](https://github.com/RobotecAI/ROSCon2024Tutorial/README.md)

Reviewer 4: ```suggestion
      > [!NOTE]  
      > If you would like to make changes to the simulation and create your own binary please follow this [README.md](https://github.com/RobotecAI/ROSCon2024Tutorial/README.md)
```

   2. Unzip it:

      ```bash
      cd rai
      # for Ubuntu 22.04 jammy and ros2 humble
      unzip -d src/examples/turtlebot4/simulation Turtlebot4_jammyhumble_0.0.1.zip

      # for Ubuntu 24.04 noble and ros2 jazzy
      unzip -d src/examples/turtlebot4/simulation Turtlebot4_noblejazzy_0.0.1.zip
      ```

3. Setup your LLM vendor: [docs/vendors.md](../../../docs/vendors.md). OpenAI or
   AWS Bedrock are recommended for models, since current local `ollama` models don't
   support vision & tool calling.
   For `embeddings_model` `ollama` vendor works well.

4. Configure `rai_whoami_node` (based on ["Your robot identity in RAI"](https://github.com/RobotecAI/rai/blob/development/docs/create_robots_whoami.md) tutorial):

> 

Reviewer 4: ```suggestion
> [!TIP]
> Generated files can be downloaded from [here](https://robotecai-my.sharepoint.com/:u:/g/personal/bartlomiej_boczek_robotec_ai/EbPZSEdXYaRGoeecu6oJg6QBsI4ZOe_mrU3uOtOflnIjQg?e=HX8ZHB) unzip them to `src/examples/turtlebot4_whoami/description/generated` with a command:
> `unzip -d src/examples/turtlebot4_whoami/description turtlebot4_whoami_generated.zip`
```

   1. Create `whoami` package for turtlebot4 in `src/examples/turtlebot4`

      ```bash
      ./scripts/create_rai_ws.sh --name turtlebot4 --destination-directory src/examples
      ```

   2. Download official turtlebot4 [data sheet](https://bit.ly/3KCp3Du) into
      `src/examples/turtlebot4_whoami/description/documentation`
   3. Download [image](https://s3.amazonaws.com/assets.clearpathrobotics.com/wp-content/uploads/2022/03/16113604/Turtlebot-4-20220207.44.png) of turtlebot4 into `src/examples/turtlebot4_whoami/description/images`
   4. Create robot's identity. Run the `parse_whoami_package`. This will process the documentation, building
      it into a vector database, which is used by RAI agent to reason about its identity.

> 
>       > **NOTE**: Be cautious if Cloud vendor was chosen step 3 as `embeddings_model`,
>       > because parsing bigger documents might lead to costs.
> 

Reviewer 4: ```suggestion
```
I think we can skip that information

>       > **NOTE**: generated files can be downloaded from [here](https://robotecai-my.sharepoint.com/:u:/g/personal/bartlomiej_boczek_robotec_ai/EbPZSEdXYaRGoeecu6oJg6QBsI4ZOe_mrU3uOtOflnIjQg?e=HX8ZHB)
>       > unzip them to `src/examples/turtlebot4_whoami/description/generated` with a command:
>       > `unzip -d src/examples/turtlebot4_whoami/description turtlebot4_whoami_generated.zip`
> 

Reviewer 4: ```suggestion
```
I have added suggestion to mention it as the first step
